iree/compiler/Codegen/Common/FlattenMemRefSubspanPass.cpp

```diff
@@ -168,39 +166,92 @@ struct FlattenBindingSubspan final
+    } else {
+      ArrayRef<int64_t> oldShape = oldType.getShape();
+      Location loc = subspanOp.getLoc();
+
+      AffineExpr sym0, sym1;
+      bindSymbols(rewriter.getContext(), sym0, sym1);
+      MLIRContext *context = rewriter.getContext();
+      auto mulMap = AffineMap::get(0, 2, {sym0 * sym1}, context);
+
+      Value totalSize = rewriter.create<ConstantIndexOp>(loc, 1);
+      int dynamicDimIndex = 0;
+      for (int i = 0; i < oldType.getRank(); ++i) {
+        Value thisDim;
+        if (ShapedType::isDynamic(oldShape[i])) {
+          thisDim = subspanOp.dynamic_dims()[dynamicDimIndex++];
+        } else {
+          thisDim = rewriter.create<ConstantIndexOp>(loc, oldShape[i]);
+        }
+        totalSize = rewriter.create<AffineApplyOp>(
+            loc, mulMap, ValueRange{totalSize, thisDim});
+      }
+      dynamicDim = totalSize;
+    }
+
     Type newType = getTypeConverter()->convertType(oldType);
 
     rewriter.replaceOpWithNewOp<IREE::HAL::InterfaceBindingSubspanOp>(
         subspanOp, newType, subspanOp.binding(), subspanOp.byte_offset(),
-        subspanOp.byte_length());
+        subspanOp.byte_length(), dynamicDim);
     return success();
   }
 };
 
 /// Generates IR to perform index linearization with the given `indices`
-/// indexing into the given memref `sourceType`.
-static Value linearizeIndices(MemRefType sourceType, ValueRange indices,
+/// indexing into the given memref `sourceValue`.
+static Value linearizeIndices(Value sourceValue, ValueRange indices,
                               Location loc, OpBuilder &builder) {
+  MemRefType sourceType = sourceValue.getType().cast<MemRefType>();
   assert(sourceType.hasRank() && sourceType.getRank() != 0);
-
-  int64_t offset;
-  SmallVector<int64_t, 4> strides;
-  if (failed(getStridesAndOffset(sourceType, strides, offset)) ||
-      llvm::is_contained(strides, MemRefType::getDynamicStrideOrOffset()) ||
-      offset == MemRefType::getDynamicStrideOrOffset()) {
-    return nullptr;
-  }
+  int64_t rank = sourceType.getRank();
 
   AffineExpr sym0, sym1, sym2;
   bindSymbols(builder.getContext(), sym0, sym1, sym2);
   MLIRContext *context = builder.getContext();
   auto mulAddMap = AffineMap::get(0, 3, {sym0 * sym1 + sym2}, context);
 
-  Value linearIndex = builder.create<ConstantIndexOp>(loc, offset);
-  for (auto pair : llvm::zip(indices, strides)) {
-    Value stride = builder.create<ConstantIndexOp>(loc, std::get<1>(pair));
+  int64_t offset;
+  SmallVector<int64_t, 4> strides;
+  strides.reserve(rank);
+
+  // First try to get the strides from the MemRef type itself. This applies to
+  // cases where we have static shapes and only the leading dimension is
+  // dynamic.
+  if (AffineMap linearLayoutMap = getStridedLinearLayoutMap(sourceType)) {
+    // Dynamic strides/offset will create symbols. There should be none for the
+    // static case.
+    if (linearLayoutMap.getNumSymbols() == 0) {
```

Comment:
I would expect that the dynamic shape will create dynamic strides and offsets. I am missing why it is done this way.

Reply:
This is for the static case. Dynamic dimensions are handled later.

Reply:
I am still not sure I follow. The case below is only handled for `ShapeCarryingInterface` ops that are only ops in IREE. We would probably need to handle the case where the symbols are dynamic "soonish". This probably the ABI side of things, but there might be memrefs within the dispatch region that also have dynamic stides and offsets. Probably with subviews. I guess this pass is run after subview folding, so maybe it doesnt matter.
Stamping it, but still not very clear what is handled and what is not.

Reply:
I see your points now. Yes it's running after FoldSubViewOps so that we have all intermediate subviews removed. Load/store take the original subspan ops as input, which has all the dynamic dim SSA values associated with them. So this works. Do you see this pattern changing "soonish"? What's coming that will break it?